Migration to dll spawner+ares+phobos

Belonit opened the pull request:
# At the moment it is still a draft

### Major changes
- **[CnCNet Spawner](https://github.com/CnCNet/yrpp-spawner)** has been completely rewritten in C++.  Fixes many minor bugs, but requires testing to make sure new ones haven't been created.
Credits: @Belonit 

- [**Ares**](https://ares.strategy-x.com/)[ v3.0p1](https://launchpad.net/ares/3.0/3.0p1) & [**Phobos**](https://github.com/Phobos-developers/Phobos) [v0.3.0.1](https://github.com/Phobos-developers/Phobos/releases/v0.3.0.1) have been added bringing a huge amount of features, bugs fixes for desyncs and other minor issues. It also brings many UI/UX improvements, like showing what's being produced for observers, power/drain on buildings (Ares), extended tooltips with time, power (and description optionally) that can be drawn not only on the sidebar, harvester counter, power surplus indicator, next idle harvester hotkey, selection priority filtering (Phobos) and much more.

### Other Features:
- Red Alert 2 and Yuri's Revenge campaign support - Credits: @Belonit, @Starkku
- Fixed many issues related to having multiple observers - Credits: @Belonit
- Singleplayer saves now working.  _For technical reasons, saves created by this game version are not compatible with the vanilla game_ - Credits: @Belonit
- Fixed an issue in RA2 mode that prevented AI players from building base defensive structures - Credits: @Belonit
- Fixed an issue in RA2 mode that caused the some animations not to show - Credits: @Belonit
- Also many other minor fixes not included in this list

## Special thanks:

@Sneer-ra2 for help with **Ranked Match Client** support

## End-User Requirements

**The XNA build additionally requires:**
* [Microsoft XNA Framework Redistributable 4.0 Refresh](https://www.microsoft.com/en-us/download/details.aspx?id=27598).

**Windows 7 SP1 and Windows 8.x additionally require:**
* Microsoft Visual C++ 2015-2019 Redistributable [64-bit](https://aka.ms/vs/16/release/vc_redist.x64.exe) and [32-bit](https://aka.ms/vs/16/release/vc_redist.x86.exe). On 64-bit systems you must install both packages.

**Windows 7 SP1 additionally requires:**
* KB3063858 [64-bit](https://www.microsoft.com/download/details.aspx?id=47442) / [32-bit](https://www.microsoft.com/download/details.aspx?id=47409).

**For Campaign support requires next file in you game folder:**
```ini
maps01.mix   ; For RA2 Allied campaign
maps02.mix   ; For RA2 Soviet campaign
mapsmd03.mix ; For Yuri's Revenge campaigns

movies01.mix ; For movies RA2 Allied campaign
movies02.mix ; For movies RA2 Soviet campaign
movmd03.mix  ; For movies Yuri's Revenge campaigns
```

**For some Russian translations of the game require:**
* In game font has been updated, there may be problems with some pirated Russian translations. If you see incorrect letters instead of Cyrillic text, add [this file](https://drive.google.com/file/d/1WKJ4Vlc-BsxkSW0uzHOJJ_BVURake_uT/view?usp=drive_link) to your game folder to fix it. This is a corrected translation from Fargus, if you need other versions of the translation - let me know

## NOTE \ TODO:
- [x] Update the Phobos to 0.4 or 0.3.1
- [ ] Release yrpp-spawner
  - not sure if needed
- [x] Resolve saves conflict between spawner and Phobos (quicksave/save game trigger action vs rewritten saves logic)
- [x] Implement Syringe DLL whitelist for main client
- [x] ditto for QM client
- [x] Implement disabling autosave in campaign
- [x] Update client with tighter AC checks
- [x] Add new hotkeys to the client
- [x] Add Phobos 0.4 hotkeys
- [ ] Enable include/inherits
- [x] Port event verification
- [x] implement and use whitelisted Syringe
- [x] [The mission is failed if the hero enters the vehicle (#349)](https://github.com/CnCNet/cncnet-yr-client-package/issues/349)
- [x] [Yuri Prime and Elite Psi-Corp in RA2 mode can control 10 instead of 1 (#296)](https://github.com/CnCNet/cncnet-yr-client-package/issues/296)
- [x] [Some animations are missing in RA2 mode (#299)](https://github.com/CnCNet/cncnet-yr-client-package/issues/299)
- [x] [Big Problem with Psy-Sensor in RA2 mode (#300)](https://github.com/CnCNet/cncnet-yr-client-package/issues/300)
- [x]  [Peculiarity with repairing the BlackHawk Transport in RA2 mode (#301)](https://github.com/CnCNet/cncnet-yr-client-package/issues/301)
- [x] [Alting behaviour in beta is incorrect (#311)](https://github.com/CnCNet/cncnet-yr-client-package/issues/311)
- [x] [Killed aircraft spawn surviving pilots (#313)](https://github.com/CnCNet/cncnet-yr-client-package/issues/313)
- [x] [RA2 Mode | Ore Purifier cash bonus on low power (#315)](https://github.com/CnCNet/cncnet-yr-client-package/issues/315)
- [x] [Killed helicopters spawn surviving pilots (#334)](https://github.com/CnCNet/cncnet-yr-client-package/issues/334)

Metadorius commented on package/Resources/ClientDefinitions.ini:
```diff
@@ -15,7 +15,7 @@ GameLauncherExecutableName=syringe.exe
 UnixGameExecutableName=Resources/Compatibility/Unix/wine-game.sh
 ; Additional command line parameters applied after -SPAWN.
-ExtraCommandLineParams=-LOG -CD
+ExtraCommandLineParams=-LOG -CD -RA2ModeSaveID=0x8d143b95
```
What is this for?

Belonit replied:
https://github.com/CnCNet/yrpp-spawner/blob/40497cdd22d27fff38b902322f23cb52e4bdf209/src/Spawner/Ra2Mode.cpp#L117

This is a way to force the spawner to switch mode to RA2 when the game is loaded. At the moment, the client does not have a mechanism for processing saves from different game modes.
I could calculate this ID dynamically, on the spawner side, however Phobos (before version 0.3.0.2) has a bug due to which it also adds its id to the mod id instead of the engine id.

Metadorius replied:
I see, so that's a temporary workaround.

Metadorius commented on game-assets/ra2mode.pack/rulesmd.ini:
```diff
@@ -19769,7 +19769,7 @@ AnimList=XGRYSML1,XGRYSML2,EXPLOSML,XGRYMED1,XGRYMED2,EXPLOMED,EXPLOLRG,TWLT070
 ; special case to only affect mechanical units
 [Mechanical]
-Verses=0%,0%,0%,100%,100%,100%,100%,100%,100%,100%,100%
+Verses=0%,0%,0%,100%,100%,100%,0%,0%,0%,100%,100%; belonit - edit needed for Ares
```
would like a comment on why the change is needed

GrantBartlett commented:
https://github.com/CnCNet/cncnet-yr-client-package/issues/313
https://github.com/CnCNet/cncnet-yr-client-package/issues/312
https://github.com/CnCNet/cncnet-yr-client-package/issues/311

Belonit commented:
I have rebased this PR to go back to the old client. For some functions, it is required to complete the[ client rollback](https://github.com/CnCNet/xna-cncnet-client/pull/494) and update it.

Belonit committed:
Delete legacy ClassicMode

Belonit committed:
Update README_Onboarding.md

Belonit committed:
Update preupdateexec

Update preupdateexec

Update preupdateexec

Belonit committed:
Update .gitattributes

Belonit committed:
Update .gitignore

Belonit committed:
Create ra2.ico

Belonit committed:
Add mix-packer script

Belonit committed:
Add sidec0*md.mix to cncnet.mix and fix lspacer.shp

Belonit committed:
Add Ares 3.0p1

Belonit committed:
Move spawner.xdp to cncnet.mix

Belonit committed:
Update licenses

Belonit committed:
Add correct vanilla loading screen bg of Observer to cncnet.mix

Belonit committed:
Put RA2 mode files in ra2mode.mix

Fix NAPSIS and NACLON graphic in RA2 Mode (#295)

Belonit committed:
Move required files from expandspawn0*.mix to cncnet.mix

Belonit committed:
Add Phobos.dll pre v0.3.1.0

0271dc3008aaf61ea1ab05c39949b7f7c504241f
https://github.com/Phobos-developers/Phobos/pull/1207

Belonit committed:
Add artmd.ini to cncnet.mix

Belonit committed:
Move spawner2.xdp to ra2mode.mix

Belonit committed:
Move seasonal crates to cncnet.mix

Belonit committed:
Move third party licenses to the Licenses folder

Belonit committed:
Update gamemd-spawn.exe

Belonit committed:
Update clienticon.ico

Belonit committed:
Add Trees Shadow Fix by Starkku

Belonit committed:
Add more files to delete to account for much older clients

Belonit committed:
Add ImprovedFont from Phobos team

https://github.com/Phobos-developers/PhobosSupplementaries/tree/develop/ImprovedFont

Belonit committed:
Update Client to 90236e4 from yr/develop (Ares build)

https://github.com/CnCNet/xna-cncnet-client/tree/90236e46e5a82205d335e623b4d970f8f7277140

Belonit committed:
Add CnCNet-Spawner.dll (CnCNetYR-Hardened-v0.0.0.6)

Belonit committed:
Fix some animations in RA2 mode (#299)

Belonit committed:
Disable DetectDisguise and SensorArray on Psychic Sensor in YR

Belonit committed:
Add support Campaign (#33 #125 #290)

Belonit committed:
Fix Ore Refinery Animation in RA2 mode

Belonit committed:
Create FHCConfig.ini

Belonit committed:
Update MainMenu.ini

Belonit committed:
Update rules for support Ares & Phobos

Belonit committed:
Update OptionsWindow.ini (#291)

Belonit committed:
Setup PlacementPreview

Belonit committed:
Fix broken AI base defence in RA2 Mode (#289)

Belonit committed:
Update ClientDefinitions.ini

Belonit committed:
Show AI on Observer sidebar

Belonit committed:
Fix an issue where IFV with Engineer could repair buildings in RA2 mode

Metadorius committed:
chmod +x all executable files

Metadorius committed:
Update spawner to v0.0.0.7

CnC-RaVaGe commented:
Bug Loading MP games.

Tried to load multiple save games that lead to a disconnect at game start.
Tried to load the earliest save file with the game crashing for both players.

[debug.20250508-115847.log](https://github.com/user-attachments/files/20101729/debug.20250508-115847.log)
[debug.log](https://github.com/user-attachments/files/20101730/debug.log)
[debug.20250508-115605.log](https://github.com/user-attachments/files/20101731/debug.20250508-115605.log)
[except.txt](https://github.com/user-attachments/files/20101735/except.txt)
[syringe.log](https://github.com/user-attachments/files/20101736/syringe.log)
[debug.log](https://github.com/user-attachments/files/20101737/debug.log)

CnC-RaVaGe commented:
Allied and Soviet UI themes must be updated to accommodate the new menus and dialogues.
Mostly missing borders so far.

Seems that lobby commands don't work in the Load Game lobby.

![Screenshot 2025-05-08 120047](https://github.com/user-attachments/assets/712ef1fe-d269-4734-ba7d-38bf14cb7cc3)


Just wondering if the loadgame lobby is visible to all or just the original participants of the game to be loaded.

SadPencil committed:
Upgrade to Client Launcher Stub 2.0.10

SadPencil committed:
Specify ClientGameType=Ares

SadPencil committed:
Update client to 09948b (v2.12.0) from yr/develop

SadPencil committed:
Delete MainMenuTheme.bak

SadPencil committed:
Update client to 09948b (v2.12.0) from yr/develop

SadPencil committed:
Update SecondStageUpdater

SadPencil committed:
Update client

Metadorius committed:
Add Ares/Phobos hotkeys

SadPencil committed:
Update client version yr/develop/beec651, based on 2.12.2, develop/0718826

SadPencil committed:
Merge branch 'develop' into feature/migration-to-dll-spawner+ares+phobos

SadPencil committed:
Merge branch 'develop' into feature/migration-to-dll-spawner+ares+phobos

SadPencil committed:
Merge branch 'develop' into feature/migration-to-dll-spawner+ares+phobos

Metadorius committed:
Implement Syringe with whitelist, fix some issues with wine launcher

Metadorius committed:
Switch to Phobos 0.4 nightly build

Metadorius committed:
Merge remote-tracking branch 'origin/develop' into feature/migration-to-dll-spawner+ares+phobos

Metadorius committed:
Update spawner to version .8

Belonit committed:
Improvements for Wine

Belonit committed:
Update YRLauncherUnix.sh

Belonit committed:
insert final newline for *.ini file

Belonit committed:
Merge branch 'develop' into feature/migration-to-dll-spawner+ares+phobos

Belonit committed:
Remove ra2-launcher.exe and ra2md-launcher.exe

Belonit committed:
Update CnCNet-Spawner.dll

to v0.0.0.9

Belonit committed:
Update Spawner and Phobos

Belonit committed:
AllowBypassBuildLimit

https://ares-developers.github.io/Ares-docs/bugfixes/type2/multipleaifactoriescloneunits.html#ai-ignores-buildlimit-when-creating-teams

CnCRAZER committed:
Extreme AI Beta changes

CnCRAZER commented:
Incase you're curious, here's my attempt of a port of the YR repo into a mode base config by Starku: 

https://github.com/CnCNet/cncnet-yr-client-package/tree/feature/mod-base-port

Curious on your thoughts on using this as a base for the migration to ares and Phobos 

@Metadorius

Metadorius commented:
> Incase you're curious, here's my attempt of a port of the YR repo into a mode base config by Starku:
> 
> https://github.com/CnCNet/cncnet-yr-client-package/tree/feature/mod-base-port
> 
> Curious on your thoughts on using this as a base for the migration to ares and Phobos
> 
> @Metadorius

could defo use it but I think we have to first release the beta (I am working on the remaining items as we speak) then on top of it use Mod Base

CnCRAZER commented:
> > Incase you're curious, here's my attempt of a port of the YR repo into a mode base config by Starku:
> 
> > 
> 
> > https://github.com/CnCNet/cncnet-yr-client-package/tree/feature/mod-base-port
> 
> > 
> 
> > Curious on your thoughts on using this as a base for the migration to ares and Phobos
> 
> > 
> 
> > @Metadorius
> 
> 
> 
> could defo use it but I think we have to first release the beta (I am working on the remaining items as we speak) then on top of it use Mod Base

Sounds good 👍🏻

Metadorius committed:
Update Phobos and spawner to test MP save improvemetns

Metadorius committed:
Update client hotkey config to Phobos 0.4 compliant one

SadPencil committed:
Update client binaries to 2.12.9

Metadorius committed:
Update Phobos to fix save trigger

Belonit committed:
update mix-packer

Metadorius committed:
Update Phobos to fix saving too fast issue

Sneer-ra2 committed:
QM V1.84d with support for new syringe launcher.

CnCRAZER committed:
Update Extreme AI Beta.ini

Remove battle bunkers prefilled with conscripts regarding soviet AI. 

Can't seem to get it working yet.

Metadorius committed:
Update Ares and Phobos with final versions of MP save amendments

Metadorius committed:
Fix quicksave hotkey description to not suggest it's SP only

Metadorius committed:
Update Phobos in attempt to fix a random crash

Metadorius committed:
Update to Phobos 0.4 Release Candidate 1

should also reduce the pathfinding crashes

CnCRAZER committed:
Restore vanilla ic behavior

IronCurtain.KeptOnDeploy=no required to keep vanilla behavior.

CnCRAZER committed:
Update Laddericon.png

Remove the ??? to match allied and soviet

CnCRAZER committed:
Update Extreme AI Beta.ini

Fix Auto Repair by removing `[IQ]` values in extreme beta file and also give Oil Derricks `ProduceCashDisplay`

CnCRAZER committed:
Update Extreme AI Beta.ini

CnCRAZER committed:
Add Invisible 0x0 Lamp Post

Introduced a new 'GALAMP' invisible light post object with 0x0 foundation for both cncnet and ra2mode packs.

Updated artmd.ini and rulesmd.ini to define the new object, and added the corresponding galamp.shp asset.

This addition enables improved mapper lighting options using Ares features.

CnCRAZER committed:
Remove lamp from buildingtypes

Metadorius committed:
Update Syringe with the fix for multithreaded hooks

CnCRAZER committed:
Add ConYard PCX Cameos + Extreme AI Beta Update

CnCRAZER committed:
pcx file relocation

CnCRAZER committed:
Update GameOptions.ini

CnCRAZER committed:
Update GameOptions.ini

Metadorius committed:
Merge develop

CnCRAZER committed:
Update updateexec

CnCRAZER committed:
fix sw for extreme ai beta

CnCRAZER committed:
Auto Save Checkboxes + Client Binary Updates

Belonit committed:
update RA2ModeSaveID for new Phobos version

CnCRAZER committed:
Fix Gunner=yes causing crashes with phobos

CnCRAZER committed:
Survival Map NOSTR additions

Also disable JP from being allied in survival gamemode

CnCRAZER committed:
Some restored TS Cameo Restorations

Metadorius committed:
Update Phobos to soon-to-be-released 0.4 RC2

Metadorius committed:
Merge branch 'develop' into feature/migration-to-dll-spawner+ares+phobos

Metadorius committed:
Update client and switch campaign autosave to `CampaignCheckBox`

Metadorius committed:
Restore old spawner `Special` house behavior

Metadorius committed:
Update Phobos with fixes to action 125

Metadorius committed:
Update spawner with improvements and support for game speed slider toggle

Metadorius committed:
Merge remote-tracking branch 'origin/develop' into feature/migration-to-dll-spawner+ares+phobos

Metadorius committed:
Merge in the changes from `artra2.ini` and `artyr.ini`

Co-authored-by: Belonit <54427022+Belonit@users.noreply.github.com>

Metadorius committed:
Direct updater to remove `artra2.ini` and `artyr.ini`

Metadorius committed:
Fix fonts interaction with translations on beta

Metadorius committed:
Adjust `RequiredFiles` for beta

Metadorius committed:
Put font file inside of a mix to workaround no default translation issue

Metadorius committed:
Fix `versionconfig.ini` to accommodate for font changes

CCCP84 committed:
Remove double definition of refinery smoke animation parameters (#661)

CCCP84 committed:
GAAIRC and AMRADR stop animation when no power (#663)

MahBoiDeveloper commented:
> 609 files changed

nice

Metadorius committed:
Merge branch 'develop' into feature/migration-to-dll-spawner+ares+phobos

Metadorius committed:
Fix and enable Ares and Phobos buttons

Metadorius committed:
Mention used projects in readme

Metadorius committed:
Add Yuri's Revenge intro description

Co-authored-by: Belonit <54427022+Belonit@users.noreply.github.com>

Metadorius committed:
Adjust tooltip on campaign autosave checkbox

Metadorius committed:
Add logos of powering projects to main menu

CCCP84 committed:
Change the positions of the powered logos (#670)

* Changing the positions of the powered logos

* Apply suggestion from @Metadorius

---------

Co-authored-by: Kerbiter <crabiter@vivaldi.net>

Metadorius committed:
Merge remote-tracking branch 'origin/develop' into feature/migration-to-dll-spawner+ares+phobos

Metadorius committed:
Adjust Syringe's license to LGPL under which the fork is produced
